### PR TITLE
EOS-26066: REST API to accept webhook information in utils rest server

### PR DESCRIPTION
**Pull request by vastradparayya**
Signed-off-by: Parayya Vastrad <parayya.vastrad@seagate.com>

# Problem Statement
- REST API to accept webhook information in utils rest server

# Design
-  https://jts.seagate.com/browse/EOS-26066

# Coding 
-  Coding conventions are followed and code is consistent [Y/N]: Y
-  Confirm All CODACY errors are resolved [Y/N]: Y

# Testing 
- [x] Confirm that Test Cases are added (for both the cases, fix and feature) [Y/N]: Y
- [x] Confirm Test Cases cover Happy Path, Non-Happy Path and Scalability [Y/N]: Y
- [x] Confirm Testing was performed with installed RPM [Y/N]:  Y

# Review Checklist 
  Before posting the PR please ensure
- [x] PR is self reviewed
- [x] Is there a change in filename/package/module or signature [Y/N]: N
- [x] If yes for above point, Is a notification sent to all other cortx components [Y/N] NA
- [x] Jira is updated
- [x] Check if the description is clear and explained. 
- [x] Check Acceptance Criterion is defined. 
- [x] All the tests performed should be mentioned before Resolving a JIRA. 
- [ ] Verification needs to be done before marked as Closed/Verified 

# Documentation
- [ ] Changes done to WIKI / Confluence page

### CODE REVIEW

File: py-utils/src/utils/audit_log/audit_log_server.py [hunk 1]
```diff
@@ -34,7 +34,29 @@ async def send(request):
         try:
             payload = await request.json()
             messages = payload['messages']
-            # TODO Store audit logs messages to a persistent storage.
+            external_server_info = payload.get('external_server_info')
+            # send audit logs messages to to external server provided by webhook
+            if not external_server_info:
+                # write the audit log message to message bus
+                from cortx.utils.message_bus import MessageProducer
+                producer = MessageProducer(producer_id='audit_log_send', \
```
Comment (sachinpunadikar) [inline py-utils/src/utils/audit_log/audit_log_server.py:42]: producer can be defined outside, so no need to define it for every send .. this producer is fixed as message_type is fixed

File: py-utils/src/utils/audit_log/audit_log_server.py [hunk 2]
```diff
@@ -34,7 +34,29 @@ async def send(request):
         try:
             payload = await request.json()
             messages = payload['messages']
-            # TODO Store audit logs messages to a persistent storage.
+            external_server_info = payload.get('external_server_info')
+            # send audit logs messages to to external server provided by webhook
+            if not external_server_info:
```
Comment (sachinpunadikar) [inline py-utils/src/utils/audit_log/audit_log_server.py:39]: Code can be optimized if you handle sending of audit message to external server first

File: py-utils/src/utils/audit_log/audit_log_server.py [hunk 3]
```diff
@@ -34,7 +34,29 @@ async def send(request):
         try:
             payload = await request.json()
             messages = payload['messages']
-            # TODO Store audit logs messages to a persistent storage.
+            external_server_info = payload.get('external_server_info')
```
Comment (sachinpunadikar) [inline py-utils/src/utils/audit_log/audit_log_server.py:37]: external_server_info is not part of the payload, its part of POST request of webhook
why you are trying get external_server_info here ?

File: py-utils/src/utils/audit_log/audit_log_server.py [hunk 4]
```diff
@@ -55,7 +77,41 @@ async def send(request):
...
+    async def send_webhook_info(request):
+        Log.debug("Received POST request for webhook information")
+        try:
+            external_server_info = await request.json()
+            # write webhook info to the external server
+            
+        except AuditLogError as e:
+            status_code = e.rc
+            error_message = e.desc
+            Log.error(f"Unable to post audit webhook information, status code: " \
```
Comment (sachinpunadikar) [inline py-utils/src/utils/audit_log/audit_log_server.py:97]: We are not POSTing .. actually we are receiving , correct the message

File: py-utils/src/utils/audit_log/audit_log_server.py [hunk 5]
```diff
@@ -55,7 +77,41 @@ async def send(request):
...
+                f"{status_code}, error: {error_message}")
+            response_obj = {'error_code': status_code, 'exception': \
+                ['AuditLogError', {'message': error_message}]}
+        except Exception as e:
+            exception_key = type(e).__name__
+            exception = RestServerError(exception_key).http_error()
+            status_code = exception[0]
+            error_message = exception[1]
+            Log.error(f"Internal error while sending webhook info." \
```
Comment (sachinpunadikar) [inline py-utils/src/utils/audit_log/audit_log_server.py:106]: We are not sending, we are receiving it

**Review by sachinpunadikar — CHANGES_REQUESTED** [review]
work on comments

**Review by rohit-k-dwivedi — COMMENTED** [review]
please work on comments and codacy issued

File: py-utils/src/utils/audit_log/audit_log_server.py (hunk 1 above)
Comment (vastradparayya) [reply py-utils/src/utils/audit_log/audit_log_server.py]: Done

File: py-utils/src/utils/audit_log/audit_log_server.py (hunk 2 above)
Comment (vastradparayya) [reply py-utils/src/utils/audit_log/audit_log_server.py]: Done

File: py-utils/src/utils/audit_log/audit_log_server.py (hunk 3 above)
Comment (vastradparayya) [reply py-utils/src/utils/audit_log/audit_log_server.py]: Done. fetched from file

File: py-utils/src/utils/audit_log/audit_log_server.py (hunk 4 above)
Comment (vastradparayya) [reply py-utils/src/utils/audit_log/audit_log_server.py]: Done

File: py-utils/src/utils/audit_log/audit_log_server.py (hunk 5 above)
Comment (vastradparayya) [reply py-utils/src/utils/audit_log/audit_log_server.py]: Done

File: py-utils/src/utils/audit_log/audit_log_server.py [hunk 6]
```diff
@@ -21,9 +21,13 @@
 from cortx.utils.utils_server import RestServer
 from cortx.utils.audit_log.error import AuditLogError
 from cortx.utils.utils_server.error import RestServerError
+from cortx.utils.message_bus import MessageProducer
+from cortx.utils.conf_store import Conf
 from cortx.utils.log import Log
 
 routes = web.RouteTableDef()
+producer = MessageProducer(producer_id='audit_log_send', \
+    message_type='audit_log_message', method='sync')
```
Comment (selvakumaar5496) [inline py-utils/src/utils/audit_log/audit_log_server.py:30]: When or Who will create this message type

**Review by selvakumaar5496 — APPROVED** [review]

File: py-utils/src/utils/audit_log/audit_log_server.py (hunk 6 above)
Comment (vastradparayya) [reply py-utils/src/utils/audit_log/audit_log_server.py:29]: audit_log_message message_type created as mini-provisioning's init phase

File: py-utils/src/setup/utils.py [hunk 7]
```diff
@@ -211,12 +211,13 @@ def post_install(post_install_template: str):
...
     def init(config_path: str):
         """ Perform initialization """
-        # Create message_type for Event Message
+        # Create message_type for Event Message and audit log 
         from cortx.utils.message_bus import MessageBusAdmin
         from cortx.utils.message_bus.error import MessageBusError
         try:
             admin = MessageBusAdmin(admin_id='register', cluster_conf=config_path)
-            admin.register_message_type(message_types=['IEM'], partitions=1)
+            admin.register_message_type(message_types=['IEM', 'audit_log_message'], \
```
Comment (sachinpunadikar) [inline py-utils/src/setup/utils.py:219]: Suggestion 
audit_log_message -> audit_messages

**Review by sachinpunadikar — APPROVED** [review]
Added a suggestion, rest all looks good

File: py-utils/src/setup/utils.py (hunk 7 above)
Comment (vastradparayya) [reply py-utils/src/setup/utils.py]: Changed message_type from audit_log_message -> audit_messages

**Review by rohit-k-dwivedi — APPROVED** [review]

**Pull request merged**